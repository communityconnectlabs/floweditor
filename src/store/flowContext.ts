import { FlowDefinition, FlowNode, UINode } from 'flowTypes';
import { combineReducers } from 'redux';
import ActionTypes, {
  UpdateAssetsAction,
  UpdateBaseLanguageAction,
  UpdateContactFieldsAction,
  UpdateDefinitionAction,
  UpdateDependenciesAction,
  UpdateNodesAction
} from 'store/actionTypes';
import Constants from 'store/constants';

// tslint:disable:no-shadowed-variable
export interface RenderNodeMap {
  [uuid: string]: RenderNode;
}

export interface RenderNode {
  ui: UINode;
  node: FlowNode;
  inboundConnections: { [nodeUUID: string]: string };
  ghost?: boolean;
}

export interface FunctionExample {
  template: string;
  output: string;
}

export interface CompletionOption {
  name?: string;
  summary: string;

  // functions
  signature?: string;
  detail?: string;
  examples?: FunctionExample[];
}

export interface ContactFields {
  [snakedFieldName: string]: string;
}

export enum AssetType {
  Channel = 'channel',
  Classifier = 'classifier',
  Contact = 'contact',
  ContactProperty = 'property',
  Currency = 'currency',
  Environment = 'environment',
  Expression = 'expression',
  Field = 'field',
  Flow = 'flow',
  Global = 'global',
  Group = 'group',
  Label = 'label',
  Language = 'language',
  Remove = 'remove',
  Resthook = 'resthook',
  Result = 'result',
  Revision = 'revision',
  Scheme = 'scheme',
  Template = 'template',
  URN = 'urn'
}

export interface Reference {
  nodeUUID: string;
  actionUUID?: string;
}

export interface Asset {
  id: string;
  name: string;
  type: AssetType;

  order?: number;
  references?: Reference[];
  isNew?: boolean;
  content?: any;
}

export const REMOVE_VALUE_ASSET = {
  id: AssetType.Remove,
  name: 'Remove Value',
  type: AssetType.Remove
};

export const DEFAULT_LANGUAGE = {
  id: 'base',
  name: 'Default',
  type: AssetType.Language
};

export interface AssetStore {
  [assetType: string]: Assets;
}

export interface AssetMap {
  [id: string]: Asset;
}

export interface Assets {
  // our local cache of assets
  items: AssetMap;

  type: AssetType;

  // an optional endpoint to search for more
  endpoint?: string;

  // option name for the id when fetching remotely
  id?: string;

  // have our assets already been loaded
  prefetched?: boolean;
}

export interface FlowContext {
  dependencies: FlowDefinition[];
  baseLanguage: Asset;
  contactFields: ContactFields;
  definition: FlowDefinition;
  nodes: { [uuid: string]: RenderNode };
  assetStore: AssetStore;
}

// Initial state
export const initialState: FlowContext = {
  definition: null,
  dependencies: null,
  baseLanguage: null,
  contactFields: {},
  nodes: {},
  assetStore: {}
};

// Action Creators
export const updateDefinition = (definition: FlowDefinition): UpdateDefinitionAction => ({
  type: Constants.UPDATE_DEFINITION,
  payload: {
    definition
  }
});

export const updateNodes = (nodes: RenderNodeMap): UpdateNodesAction => ({
  type: Constants.UPDATE_NODES,
  payload: {
    nodes
  }
});

export const updateDependencies = (dependencies: FlowDefinition[]): UpdateDependenciesAction => ({
  type: Constants.UPDATE_DEPENDENCIES,
  payload: {
    dependencies
  }
});

export const updateBaseLanguage = (baseLanguage: Asset): UpdateBaseLanguageAction => ({
  type: Constants.UPDATE_BASE_LANGUAGE,
  payload: {
    baseLanguage
  }
});

export const updateContactFields = (contactFields: ContactFields): UpdateContactFieldsAction => ({
  type: Constants.UPDATE_CONTACT_FIELDS,
  payload: {
    contactFields
  }
});

export const updateAssets = (assets: AssetStore): UpdateAssetsAction => ({
  type: Constants.UPDATE_ASSET_MAP,
  payload: {
    assets
  }
});

// Reducers
export const definition = (
  state: FlowDefinition = initialState.definition,
  action: ActionTypes
) => {
  switch (action.type) {
    case Constants.UPDATE_DEFINITION:
      return action.payload.definition;
    default:
      return state;
  }
};

export const nodes = (state: {} = initialState.nodes, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_NODES:
      return action.payload.nodes;
    default:
      return state;
  }
};

export const dependencies = (
  state: FlowDefinition[] = initialState.dependencies,
  action: ActionTypes
) => {
  switch (action.type) {
    case Constants.UPDATE_DEPENDENCIES:
      return action.payload.dependencies;
    default:
      return state;
  }
};

export const assetStore = (state: AssetStore = initialState.assetStore, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_ASSET_MAP:
      return action.payload.assets;
    default:
      return state;
  }
};

export const baseLanguage = (state: Asset = initialState.baseLanguage, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_BASE_LANGUAGE:
      return action.payload.baseLanguage;
    default:
      return state;
  }
};

export const contactFields = (
  state: ContactFields = initialState.contactFields,
  action: ActionTypes
) => {
  switch (action.type) {
    case Constants.UPDATE_CONTACT_FIELDS:
      return action.payload.contactFields;
    default:
      return state;
  }
};

// Root reducer
export default combineReducers({
  definition,
  nodes,
  dependencies,
  assetStore,
  baseLanguage,
  contactFields
});
