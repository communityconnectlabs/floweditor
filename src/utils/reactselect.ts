import { StylesConfig } from 'react-select';

export const large = {
  placeholder: (styles: StylesConfig, state: any) => {
    return { ...styles, color: '#ddd', marginLeft: '3px' };
  },
  singleValue: (styles: StylesConfig, state: any) => {
    return { ...styles, marginLeft: '3px' };
  },
  multiValue: (styles: StylesConfig, state: any) => {
    return { ...styles, marginLeft: '0px' };
  },
  input: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      marginLeft: '0px',
      caretColor: '#999',
      marginBottom: '0px',
      boxShadow: 'none',
      border: 'none'
    };
  },
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#ccc',
      boxShadow: 'none',
      transition: 'all 0.3s ease-in-out',
      '&:hover': {
        borderColor: '#ccc'
      }
    };
  }
};

export const messageStyle = {
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#ccc',
      boxShadow: 'none',
      borderBottom: 'none',
      borderBottomRightRadius: '0px',
      borderBottomLeftRadius: '0px',
      '&:hover': {
        borderColor: '#ccc'
      }
    };
  }
};

export const getErroredSelect = (baseControl: any): any => {
  return {
    control: (styles: StylesConfig) => ({
      ...styles,
      ...baseControl,
      borderColor: 'tomato',
      boxShadow: '0 0 0 3px rgba(255,196,186,0.5) !important;',
      '&:hover': {
        borderColor: 'tomato'
      }
    })
  };
};

export const small = {
  control: (styles: StylesConfig) => ({
    ...styles,
    borderColor: '#ccc',
    boxShadow: 'none',
    '&:hover': {
      borderColor: '#ccc'
    },
    maxHeight: '28px',
    minHeight: '28px',
    fontSize: '12px',
    lineHeight: '12px'
  }),
  menuList: (styles: StylesConfig) => ({
    ...styles,
    maxHeight: 200
  }),
  option: (styles: StylesConfig) => ({
    ...styles,
    fontSize: 12
  }),
  singleValue: (styles: StylesConfig) => ({
    ...styles,
    maxHeight: 28,
    lineHeight: '13px'
  }),
  indicatorsContainer: (styles: StylesConfig) => ({
    ...styles,
    height: 28,
    maxHeight: 28
  }),
  valueContainer: (styles: StylesConfig) => ({
    ...styles,
    height: 28,
    maxHeight: 28
  }),
  input: (styles: StylesConfig) => ({
    ...styles,
    marginLeft: '-2px',
    marginTop: '-10px',
    caretColor: '#999',
    boxShadow: 'none',
    border: 'none',
    maxHeight: '20px',
    minHeight: '20px',
    lineHeight: '12px'
  })
};

export const tagging = {
  control: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      borderColor: '#ccc',
      boxShadow: 'none',
      transition: 'all 0.3s ease-in-out',
      '&:hover': {
        borderColor: '#ccc'
      }
    };
  },
  input: (styles: StylesConfig, state: any) => {
    return {
      ...styles,
      marginLeft: '0px',
      caretColor: '#999',
      marginBottom: '0px',
      boxShadow: 'none',
      border: 'none'
    };
  },
  multiValue: (styles: StylesConfig) => {
    return {
      ...styles,
      maxHeight: 24,
      minHeight: 24,
      lineHeight: '17px',
      backgroundColor: 'hsl(0, 0%, 94%)',
      border: '1px solid hsl(0, 0%, 85%)'
    };
  }
};
