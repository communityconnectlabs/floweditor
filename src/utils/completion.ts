import { AssetStore, CompletionOption } from 'store/flowContext';

export interface CompletionAssets {
  assetStore: AssetStore;
  schema: CompletionSchema;
}

export interface CompletionProperty {
  key: string;
  help: string;
  type: string;
}

export interface CompletionType {
  name: string;

  key_source?: string;
  property_template?: CompletionProperty;
  properties?: CompletionProperty[];
}

export interface CompletionSchema {
  types: CompletionType[];
  root: CompletionProperty[];
}

export const getFunctions = (functions: CompletionOption[], query: string): CompletionOption[] => {
  return functions.filter((option: CompletionOption) => {
    if (option.signature) {
      return option.signature.indexOf(query) === 0;
    }
    return false;
  });
};

/**
 * Takes a dot query and returns the completions options at the current level
 * @param dotQuery query such as "results.field_name.ca"
 */
export const getCompletions = (
  completion: CompletionAssets,
  dotQuery: string
): CompletionOption[] => {
  const parts = (dotQuery || '').split('.');
  let currentProps: CompletionProperty[] = completion.schema.root;

  let prefix = '';
  let part = '';
  while (parts.length > 0) {
    part = parts.shift();
    if (part) {
      // eslint-disable-next-line
      const nextProp = currentProps.find((prop: CompletionProperty) => prop.key === part);
      if (nextProp) {
        // eslint-disable-next-line
        const nextType = completion.schema.types.find(
          (type: CompletionType) => type.name === nextProp.type
        );
        if (nextType && nextType.properties) {
          currentProps = nextType.properties;
          prefix += part + '.';
        } else if (nextType && nextType.property_template) {
          prefix += part + '.';
          const template = nextType.property_template;
          currentProps = Object.keys(completion.assetStore[nextType.name].items).map(
            (key: string) => ({
              key: template.key.replace('{key}', key),
              help: template.help.replace('{key}', key),
              type: template.type
            })
          );
        } else {
          // eslint-disable-next-line
          currentProps = currentProps.filter((prop: CompletionProperty) =>
            prop.key.startsWith(part.toLowerCase())
          );
          break;
        }
      } else {
        // eslint-disable-next-line
        currentProps = currentProps.filter((prop: CompletionProperty) =>
          prop.key.startsWith(part.toLowerCase())
        );
        break;
      }
    }
  }

  return currentProps.map((prop: CompletionProperty) => {
    const name =
      prop.key === '__default__' ? prefix.substr(0, prefix.length - 1) : prefix + prop.key;
    return { name, summary: prop.help };
  });
};

export const getCompletionName = (option: CompletionOption): string => {
  return option.name || option.signature.substr(0, option.signature.indexOf('('));
};

export const getCompletionSignature = (option: CompletionOption): string => {
  return option.signature.substr(option.signature.indexOf('('));
};
