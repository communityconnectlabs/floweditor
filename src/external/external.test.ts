import { resultToAsset } from './index';
import { AssetType } from '../store/flowContext';
import * as utils from '../utils';
import { mock } from 'testUtils';

mock(utils, 'createUUID', utils.seededUUIDs());
describe('External', () => {
  it('result type should trump request type', () => {
    const group_uuid = utils.createUUID();
    const group = resultToAsset(
      { uuid: group_uuid, name: 'My Test Group', type: AssetType.Group },
      AssetType.Contact,
      null
    );

    expect(group.type).toEqual(AssetType.Group);
    expect(group).toMatchSnapshot();
  });

  it('converts rp api flow types', () => {
    // survey converts to the goflowy message_offline
    expect(resultToAsset({ type: 'survey' }, AssetType.Flow, 'uuid').content.type).toEqual(
      'message_offline'
    );

    // message and voice stay the same
    expect(resultToAsset({ type: 'message' }, AssetType.Flow, 'uuid').content.type).toEqual(
      'message'
    );

    expect(resultToAsset({ type: 'voice' }, AssetType.Flow, 'uuid').content.type).toEqual('voice');
  });
});
