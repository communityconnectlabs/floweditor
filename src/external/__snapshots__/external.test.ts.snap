// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`External result type should trump request type 1`] = `
Object {
  "content": Object {
    "type": "group",
  },
  "id": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
  "name": "My Test Group",
  "type": "group",
}
`;
